### PR TITLE
Torchhub: automatically disambiguate ref if it's both a branch and a tag

### PR DESCRIPTION
**Pull request by NicolasHug**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74418

Differential Revision: [D35008926](https://our.internmc.facebook.com/intern/diff/D35008926)

### CODE REVIEW

File: torch/hub.py [hunk 1]
```diff
@@ -111,32 +112,32 @@ def _get_torch_home():
 
 def _parse_repo_info(github):
     if ':' in github:
-        repo_info, branch = github.split(':')
+        repo_info, ref = github.split(':')
```
Comment (NicolasHug) [inline torch/hub.py:115]: I renamed (private) uses of the term "branch" into "ref", as this is more accurate. This is somewhat orthogonal to the relevant changes made in this PR, so I can revert if it makes the review difficult.

**Review by NicolasHug — COMMENTED** [review]
Closes https://github.com/pytorch/pytorch/issues/74368

This PR does 2 things:

- Update the URL where we download the archive, according to the latest github API docs https://docs.github.com/en/rest/reference/repos#download-a-repository-archive-zip
- When a user hits a 300 because the requested ref exists both as a branch and a tag, torchhub now re-tries to download the repo, assuming it's a branch (just like `git`). 

I'm not adding test right now because the tests require access to Ailing's private repo. We should find a better way to do this. But meanwhile, the following testplan will be enough:

```py
from pytest import skip
import torch

try:
    torch.hub.load("pytorch/vision:v0.5.0", "resnet18", force_reload=True)
except ImportError as e:
    # We're loading a super old torchvision version so the import error is expected.
    # It's completely unrelated to the changes made to torchhub.
    assert "cannot import name 'PY3' from 'torch._six'" in str(e)
    pass

# We can also load tags and branches explicitly without issue:
torch.hub.load("pytorch/vision:refs/tags/v0.5.0", "resnet18", force_reload=True, skip_validation=True)
torch.hub.load("pytorch/vision:refs/heads/v0.5.0", "resnet18", force_reload=True, skip_validation=True)
```

**Review by vmoens — APPROVED** [review]
LGTM thanks for fixing this!

**Comment by NicolasHug** [comment]
@NicolasHug has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35008926).

**Pull request closed without merging**